pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1870 @@
+        pub fn get_collator_stakable_free_balance(acc: &T::AccountId) -> BalanceOf<T> {
+            let mut balance = <T as Config>::Currency::free_balance(acc);
+            if let Some(info) = <CandidateInfo<T>>::get(acc) {
+                balance = balance.saturating_sub(info.bond);
+            }
+            balance
+        }
+        /// One-off fn that initializes pallet parameters to onboard the pallet after genesis
+        pub fn initialize_pallet(
+            starting_block: <T as frame_system::Config>::BlockNumber,
+            candidates: Vec<T::AccountId>,
+            inflation: InflationInfo<BalanceOf<T>>,
+        ) -> Result<(), Error<T>> {
+            ensure!(
+                Self::on_chain_storage_version() == 0,
+                <Error<T>>::PalletAlreadyInitialized
+            );
+
+            // Ensure we have at least MinSelected candidates in the list
+            ensure!(
+                candidates.len() >= T::MinSelectedCandidates::get() as usize,
+                <Error<T>>::NotEnoughCollatorsForPalletInit
+            );
+
+            // provided by https://docs.rs/frame-support/latest/frame_support/traits/trait.OnGenesis.html
+            // Sets STORAGE_VERSION
+            <Self as frame_support::traits::OnGenesis>::on_genesis();
+            assert_eq!(
+                Self::current_storage_version(),
+                Self::on_chain_storage_version(),
+            );
+
+            <InflationConfig<T>>::put(inflation);
+
+            let whitelist_bond = T::MinWhitelistCandidateStk::get();
+
+            let mut candidate_count = 0u32;
+            // Initialize the candidates
+            for candidate in candidates {
+                ensure!(
+                    <T as Config>::Currency::reserved_balance(&candidate) >= whitelist_bond,
+                    <Error<T>>::InsufficientBalance
+                );
+                if let Err(error) = <Pallet<T>>::join_candidates(
+                    T::Origin::from(Some(candidate.clone()).into()),
+                    whitelist_bond,
+                    candidate_count,
+                ) {
+                    log::warn!("Join candidates failed in migration with error {:?}", error);
+                } else {
+                    candidate_count = candidate_count.saturating_add(1u32);
+                }
+            }
+            log::info!("Added {} initial candidates", candidate_count);
+
+            // Set collator commission to default config
+            <CollatorCommission<T>>::put(T::DefaultCollatorCommission::get());
+            // Set parachain bond config to default config
+            <ParachainBondInfo<T>>::put(ParachainBondConfig {
+                // must be set soon; if not => due inflation will be sent to collators/delegators
+                account: T::AccountId::decode(&mut sp_runtime::traits::TrailingZeroInput::zeroes())
+                    .expect("infinite length input; no invalid inputs for type; qed"),
```

Comment:
expect

Reply:
expect makes sense here, this can't panic however long the AccountId concrete type is